File: packages/javascript/lib/_internal/parse.ts
// Helpers that keep the datasets readable: pools are written as whitespace-
// separated strings inside a template literal instead of one array entry per
// line, which keeps a 120-name pool to a handful of lines.

import type { PredicateForm, PredicateTense } from '../sentence/data/types.js';
import type { WordGender, WordPool } from '../word/data/types.js';

/**
 * Split a whitespace-separated pool. `_` stands for a space inside a single
 * entry, so multi-word names survive the split (`De_Luca` -> `De Luca`).
 */
export function words(source: string): readonly string[] {
	return source
		.trim()
		.split(/\s+/)
		.map((word) => word.replace(/_/g, ' '));
}

/**
 * Split a `theme -> \`gato:m luna:f\`` map into the pools and the gender lookup
 * the two of them share. One pass over one source, so a language that inflects
 * still writes each noun exactly once.
 */
export function taggedNouns<T extends string>(
	source: Record<T, string>
): { pools: Record<T, WordPool>; gender: Record<string, WordGender> } {
	const pools = {} as Record<T, WordPool>;
	const gender: Record<string, WordGender> = {};

	for (const theme of Object.keys(source) as T[]) {
		pools[theme] = words(source[theme]).map((entry) => {
			const at = entry.lastIndexOf(':');
			const word = entry.slice(0, at);

			gender[word] = entry.slice(at + 1) as WordGender;

			return word;
		});
	}

	return { pools, gender };
}

/**
 * Split a whitespace-separated pool of `native:roman` pairs, for scripts whose
 * characters carry their own reading (Japanese kanji, Chinese hanzi).
 */
export function tokens(source: string): readonly { n: string; r: string }[] {
	return words(source).map((pair) => {
		const [n, r] = pair.split(':');
		return { n, r };
	});
}

/**
 * Split a whitespace-separated pool of `native:weight` pairs into a lookup, for
 * pools whose entries are not equally likely (surname frequency). Entries left
 * out of the source keep whatever default the caller falls back to.
 */
export function weights(source: string): Record<string, number> {
	const map: Record<string, number> = {};

	for (const { n, r } of tokens(source)) {
		map[n] = Number(r);
	}

	return map;
}

/** Build a native -> romanization lookup from `native:roman` pairs. */
export function romanMap(source: string): Record<string, string> {
	const map: Record<string, string> = {};

	for (const { n, r } of tokens(source)) {
		map[n] = r;
	}

	return map;
}

/**
 * Every form of a tense from one pool of stems and one ending per form, for a
 * language whose endings are the same whatever the stem: a Korean past stem
 * closes on `ㅆ`, so `달렸` takes `다`, `니`, `구나`, `어요` and `습니다` exactly
 * the way `걸었` does. An ending may list alternatives with `|` between them, and
 * each stem gets every one of them, so the pools stay index-aligned with the
 * present-tense `words` the stems were written for.
 */
export function conjugate(
	stems: string,
	endings: { statement: string } & Partial<Record<PredicateForm, string>>
): PredicateTense {
	const bases = words(stems);
	const attach = (ending: string): readonly string[] =>
		bases.map((stem) =>
			ending
				.split('|')
				.map((each) => stem + each)
				.join('|')
		);
	const forms: Partial<Record<PredicateForm, readonly string[]>> = {};

	for (const key of Object.keys(endings) as (keyof typeof endings)[]) {
		if (key !== 'statement') {
			forms[key] = attach(endings[key]!);
		}
	}

	return { words: attach(endings.statement), forms };
}

/** One division an outline names, and where it sits. */
export interface OutlineEntry {
	/**
	 * The division's name and the name of every division it sits inside, largest
	 * first, one per level down to its own. `null` for a level its branch skips.
	 */
	readonly path: readonly (string | null)[];
	/** The level the division itself is, as an index into the dataset's levels. */
	readonly depth: number;
	/** The shallowest level among the divisions directly inside it, or `null` for none. */
	readonly below: number | null;
}

/**
 * Split an outline of divisions, `levels` deep. A line `# name` opens a division
 * at the first level and `## name` one at the second; a line with no marker is a
 * pool of divisions at the last level, inside the division opened last. `_`
 * stands for a space, the way it does in `words`.
 *
 * A pool straight after a `#` line skips the levels between: 세종특별자치시 has no
 * 시·군·구, so its 읍·면·동 follow its own line.
 */
export function outline(source: string, levels: number): OutlineEntry[] {
	type Open = { path: (string | null)[]; depth: number; below: number | null };

	const entries: Open[] = [];
	// The division opened last at each level, down to the deepest one still open.
	const open: Open[] = [];

	const add = (name: string, depth: number, parent: Open | undefined) => {
		const path = parent ? [...parent.path] : [];

		while (path.length < depth) {
			path.push(null);
		}

		path.push(name.replace(/_/g, ' '));

		const entry: Open = { path, depth, below: null };

		if (parent) {
			parent.below = parent.below === null ? depth : Math.min(parent.below, depth);
		}

		entries.push(entry);

		return entry;
	};

	for (const line of source.split('\n')) {
		const text = line.trim();

		if (!text) {
			continue;
		}

		const marker = /^(#+) (\S+)$/.exec(text);

		if (marker) {
			const depth = marker[1].length - 1;

			open.length = depth;
			open.push(add(marker[2], depth, open[depth - 1]));
			continue;
		}

		for (const name of text.split(/\s+/)) {
			add(name, levels - 1, open[open.length - 1]);
		}
	}

	return entries;
}

File: packages/python/src/randino/_internal/parse.py
"""Helpers that keep the datasets readable.

Pools are written as whitespace-separated strings inside a triple-quoted string
instead of one list entry per line, which keeps a 120-name pool to a handful of
lines.
"""

import re
from collections.abc import Mapping
from typing import TYPE_CHECKING, NamedTuple, cast

from randino._types import WordTheme
from randino.word.data._types import WordGender, WordPool

if TYPE_CHECKING:
    from randino.sentence.data._types import PredicateTense


class NameToken(NamedTuple):
    """A name part that carries its own romanization (Japanese kanji, Chinese hanzi)."""

    n: str
    """The part in its own script."""

    r: str
    """How that part is read in the Latin alphabet."""


def words(source: str) -> tuple[str, ...]:
    """Split a whitespace-separated pool.

    `_` stands for a space inside a single entry, so multi-word names survive the
    split (`De_Luca` -> `De Luca`).
    """
    return tuple(word.replace("_", " ") for word in source.split())


def tokens(source: str) -> tuple[NameToken, ...]:
    """Split a whitespace-separated pool of `native:roman` pairs.

    For scripts whose characters carry their own reading (Japanese kanji, Chinese
    hanzi).
    """
    return tuple(NameToken(*pair.split(":", 1)) for pair in words(source))


def weights(source: str) -> dict[str, int]:
    """Split a whitespace-separated pool of `native:weight` pairs into a lookup.

    For pools whose entries are not equally likely (surname frequency). Entries left
    out of the source keep whatever default the caller falls back to.
    """
    return {token.n: int(token.r) for token in tokens(source)}


def roman_map(source: str) -> dict[str, str]:
    """Build a native -> romanization lookup from `native:roman` pairs."""
    return {token.n: token.r for token in tokens(source)}


def tagged_nouns(
    source: Mapping[WordTheme, str],
) -> tuple[dict[WordTheme, WordPool], dict[str, WordGender]]:
    """Split a `theme -> "gato:m luna:f"` map into the pools and the gender lookup.

    One pass over one source, so a language that inflects still writes each noun
    exactly once.
    """
    pools: dict[WordTheme, WordPool] = {}
    gender: dict[str, WordGender] = {}

    for theme, pool in source.items():
        entries = []

        for tagged in words(pool):
            word, _, tag = tagged.rpartition(":")

            gender[word] = cast(WordGender, tag)
            entries.append(word)

        pools[theme] = tuple(entries)

    return pools, gender


def conjugate(stems: str, endings: Mapping[str, str]) -> "PredicateTense":
    """Every form of a tense from one pool of stems and one ending per form.

    For a language whose endings are the same whatever the stem: a Korean past stem
    closes on `ㅆ`, so `달렸` takes `다`, `니`, `구나`, `어요` and `습니다` exactly the
    way `걸었` does. `endings` maps `"statement"` and any form to its ending; an ending
    may list alternatives with `|` between them, and each stem gets every one of them, so
    the pools stay index-aligned with the present-tense words the stems were written for.

    Args:
        stems: The past stems, whitespace-separated.
        endings: The ending each form takes, `"statement"` included.

    Returns:
        The tense, with the statement in `words` and the rest in `forms`.
    """
    bases = words(stems)

    def attach(ending: str) -> tuple[str, ...]:
        return tuple("|".join(stem + each for each in ending.split("|")) for stem in bases)

    # Imported here rather than at the top: the sentence package imports this module
    # through its data files, and a module-level import would be a cycle.
    from randino.sentence.data._types import PredicateForm, PredicateTense

    return PredicateTense(
        words=attach(endings["statement"]),
        forms={
            cast(PredicateForm, form): attach(ending)
            for form, ending in endings.items()
            if form != "statement"
        },
    )


class OutlineEntry(NamedTuple):
    """One division an outline names, and where it sits."""

    path: tuple[str | None, ...]
    """The division's name and the name of every division it sits inside.

    Largest first, one per level down to its own. None for a level its branch skips.
    """

    depth: int
    """The level the division itself is, as an index into the dataset's levels."""

    below: int | None
    """The shallowest level among the divisions directly inside it, or None for none."""


_MARKER = re.compile(r"(#+) (\S+)")
"""A line that opens a division: one `#` per level down, then the division's name."""


def outline(source: str, levels: int) -> tuple[OutlineEntry, ...]:
    """Split an outline of divisions, `levels` deep.

    A line `# name` opens a division at the first level and `## name` one at the second;
    a line with no marker is a pool of divisions at the last level, inside the division
    opened last. `_` stands for a space, the way it does in `words`.

    A pool straight after a `#` line skips the levels between: 세종특별자치시 has no
    시·군·구, so its 읍·면·동 follow its own line.

    Args:
        source: The outline, one division or one pool of them per line.
        levels: How many levels the outline holds.

    Returns:
        Every division the outline names, in the order it names them.
    """
    # Built as three parallel lists and frozen at the end, because a division learns
    # what is `below` it only once the lines inside it have been read.
    paths: list[tuple[str | None, ...]] = []
    depths: list[int] = []
    belows: list[int | None] = []
    # The division opened last at each level, as an index into the lists above, down to
    # the deepest one still open. None where a marker skipped a level.
    opened: list[int | None] = []

    def add(name: str, depth: int, parent: int | None) -> int:
        path: list[str | None] = [] if parent is None else list(paths[parent])

        while len(path) < depth:
            path.append(None)

        path.append(name.replace("_", " "))

        if parent is not None:
            below = belows[parent]
            belows[parent] = depth if below is None else min(below, depth)

        paths.append(tuple(path))
        depths.append(depth)
        belows.append(None)

        return len(paths) - 1

    for line in source.split("\n"):
        text = line.strip()

        if not text:
            continue

        marker = _MARKER.fullmatch(text)

        if marker:
            depth = len(marker.group(1)) - 1

            del opened[depth:]

            while len(opened) < depth:
                opened.append(None)

            parent = opened[depth - 1] if depth > 0 else None
            opened.append(add(marker.group(2), depth, parent))
            continue

        parent = opened[-1] if opened else None

        for name in text.split():
            add(name, levels - 1, parent)

    return tuple(
        OutlineEntry(path, depth, below)
        for path, depth, below in zip(paths, depths, belows, strict=True)
    )
